### PR TITLE
Domain model improvements

### PR DESCRIPTION
**Pull request by vidurananayakkara**
Domain prioritising and other improvements

### CODE REVIEW

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityComponent.java [hunk 1]
```diff
@@ -325,137 +327,130 @@ public void onAllRequiredCapabilitiesAvailable() {
...
-
-                    if (null != identityStoreConnectorConfig) {
-                        String connectorType = identityStoreConnectorConfig.getConnectorType();
-
-                        IdentityStoreConnectorFactory identityStoreConnectorFactory = CarbonSecurityDataHolder
-                                .getInstance().getIdentityStoreConnectorFactoryMap().get(connectorType);
+            DomainConfig domainConfig = DomainConfigBuilder.getDomainConfig(claimConfig.getMetaClaims());
+            CarbonSecurityDataHolder.getInstance().setDomainConfig(domainConfig);
+        } catch (DomainException e) {
+            log.error(e.getMessage(), e);
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityComponent.java]: Add a custom error message instead of e.getMessage()

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityComponent.java [hunk 2]
```diff
@@ -325,137 +327,130 @@ public void onAllRequiredCapabilitiesAvailable() {
...
+            credentialStore.init(domainManager, storeConfig.getCredentialConnectorConfigMap());
+            identityStore.init(domainManager, storeConfig.getIdentityConnectorConfigMap());
+            authorizationStore.init(storeConfig.getAuthorizationConnectorConfigMap());
+        } catch (CredentialStoreException | AuthorizationStoreException | IdentityStoreException e) {
+            log.error(e.getMessage(), e);
+        }
 
-                domain.setClaimMappings(domainMetaClaimMappings);
+        // Register the carbon realm service.
+        // TODO: Validate the configuration files for multiple primary attributes.
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityComponent.java]: Is this TODO still valid?

Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityComponent.java]: There will not be multiple primary attributes. Therefore this TODO should be removed

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/config/DomainConfigBuilder.java [hunk 3]
```diff
@@ -93,17 +96,32 @@ public static DomainConfig getDomainConfig(Map<String, MetaClaim> metaClaims) {
                 identityStoreConnectorMapping.put(identityStoreConnectorId, metaClaimMappings);
             }
 
+            // Check for duplicate domain names
+            // It is not necessary to check other maps since the same domain name is added
+            if (domainIdentityStoreConnectors.containsKey(domainName)) {
+                throw new DomainException(String
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/config/DomainConfigBuilder.java]: This should be a ConfigException not a DomainException.

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java [hunk 4]
```diff
@@ -18,95 +18,226 @@
...
+    }
+
+    /**
+     * Get all the domains to a domain name -> domain instance map.
+     *
+     * @return Map<String, Domain> map between domain name to Domain
+     */
+    private Map<String, Domain> getAllDomainNameToDomainMappingMap() {
+
+        Map<String, Domain> domainMap = new HashMap<>();
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java]: Instead of creating a new Map for each call this map should be pre-populated and saved as a class attribute.

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java [hunk 5]
```diff
@@ -18,95 +18,226 @@
...
+        // relevant domain. If no domain found throw an exception
+        Map<String, Domain> domainNameToDomainMap = getAllDomainNameToDomainMappingMap();
 
-        return getDomainFromName(CarbonSecurityConstants.DEFAULT_DOMAIN_NAME);
+        for (Domain domain : domainNameToDomainMap.values()) {
+
+            Map<String, IdentityStoreConnector> identityStoreConnectorsMap =
+                    domain.getIdentityStoreConnectorMap();
+
+            for (IdentityStoreConnector identityStoreConnector : identityStoreConnectorsMap.values()) {
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java]: Looks like username is not taken into consideration here?

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java [hunk 6]
```diff
@@ -18,95 +18,226 @@
...
 
-    // TODO <VIDURA> Add implementation
     @Override
-    public Domain getDomainFromUserName(String username) {
-        return null;
+    public void addIdentityStoreConnectorToDomain(
+            IdentityStoreConnector identityStoreConnector,
+            String domainName) throws DomainException {
+
+        Domain domain = this.getDomainFromDomainName(domainName);
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java]: Do we need this keyword when calling a method?

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java [hunk 7]
```diff
@@ -18,95 +18,226 @@
...
+    }
+
+    @Override
+    public Domain getDomainFromClaim(Claim claim) throws DomainException {
+
+        Set<Map.Entry<Integer, Map<String, Domain>>> set = domainPriorityToDomainMap.entrySet();
+
+        for (Map.Entry<Integer, Map<String, Domain>> entry : set) {
+
+            for (Map.Entry<String, Domain> domainMapEntry : entry.getValue().entrySet()) {
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java]: Since key of this entry is not used we can iterate through entry values. Also check if we can avoid having a map inside a map.

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CacheBackedIdentityStore.java [hunk 8]
```diff
@@ -86,7 +87,8 @@ public void init(DomainManager domainManager, Map<String, IdentityStoreConnector
     }
 
     @Override
-    public User getUser(String username) throws IdentityStoreException, UserNotFoundException {
+    public User getUser(String username)
+            throws IdentityStoreException, UserNotFoundException, DomainException {
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CacheBackedIdentityStore.java]: Domain exception should be wrapped from IdentityStoreException since the caller does not have any value in knowing them differently.

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java [hunk 9]
```diff
@@ -127,14 +120,17 @@ public AuthenticationContext authenticate(Callback[] callbacks) throws Authentic
...
-            credentialStoreConnectorsMap = this.domainManager.getDefaultDomain().getCredentialStoreConnectorMap();
-        } catch (DomainManagerException e) {
+            credentialStoreConnectorsMap = resolveDomain(callbacks).getCredentialStoreConnectorMap();
+        } catch (CredentialStoreException e) {
             credentialStoreConnectorsMap = Collections.emptyMap();
+            log.error(e.getMessage());
         }
 
+        AuthenticationFailure authenticationFailure = new AuthenticationFailure("Invalid user credentials.");
+
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java]: Why do we create this object even before catching an exception?

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java [hunk 10]
```diff
@@ -108,46 +99,45 @@ public User getUser(String username) throws IdentityStoreException, UserNotFound
...
         String attributeName = claim.getClaimURI(); // TODO: Get the attribute name from the claim manager.
         String attributeValue = claim.getValue();
 
-        Map<String, IdentityStoreConnector> identityStoreConnectorsMap = resolveDomain().getIdentityStoreConnectorMap();
+        Map<String, IdentityStoreConnector> identityStoreConnectorsMap = resolveDomain(claim)
+                .getIdentityStoreConnectorMap();
 
         for (IdentityStoreConnector identityStoreConnector : identityStoreConnectorsMap.values()) {
             try {
+                // The getUserBuilder method will set a unique user id
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java]: getUserBuilder belongs to a connector. It only has the connector specific user Id. Unique userId should be set by the IdentityStore.

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java [hunk 11]
```diff
@@ -217,31 +214,46 @@ public User getUser(Callback[] callbacks) throws IdentityStoreException, UserNot
...
-        // TODO: Check for domain information in userId
-        Map<String, IdentityStoreConnector> identityStoreConnectorsMap = resolveDomain().getIdentityStoreConnectorMap();
+        Map<String, IdentityStoreConnector> identityStoreConnectorsMap;
+
+        try {
+            identityStoreConnectorsMap = this.domainManager
+                    .getDomainFromUserName(userName).getIdentityStoreConnectorMap();
+        } catch (DomainException e) {
+            IdentityStoreException identityStoreException = new IdentityStoreException();
+            identityStoreException.addSuppressed(e);
```
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java]: Why is this added as a suppressed exception?

**Comment by akalankapagoda** [comment]
Can we rebase and merge these commits into one as well.

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java (hunk 9 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java]: This is required since the CredentialStoreExceptions are suppressed to the AuthenticationFailure and is thrown after the loop finished

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java (hunk 4 above)
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java]: In this case, we are maintaining Map<String, Domain> and a TreeMap<Integer, Map<String, Domain> which improves performance well.

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStore.java [hunk 12]
```diff
@@ -184,17 +185,10 @@ void init(DomainManager domainManager, Map<String, IdentityStoreConnectorConfig>
...
      */
-    boolean isUserInGroup(String userId, String groupId) throws IdentityStoreException;
-
-    /**
-     * Get all available identity store connector names.
-     *
-     * @return Map of display names against the connector id.
-     */
-    Map<String, String> getAllIdentityStoreNames();
+    boolean isUserInGroup(String userName, String groupId) throws IdentityStoreException;
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStore.java]: Users in identity store are resolved using a unique / primary attribute. This will be added as a TODO

**Comment by vidurananayakkara** [comment]
Do verify the changes made in the pull request. Upon completed verification, a new pull request will be opened from a different branch discarding this pull request which will contain only a single commit

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java (hunk 4 above)
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/domain/InMemoryDomainManager.java]: How is creating a new map in the middle of the processing going to improve performance?

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java (hunk 9 above)
Comment (akalankapagoda) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java]: I don't think we need to suppress the below exception.

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityComponent.java [hunk 13]
```diff
@@ -325,137 +327,130 @@ public void onAllRequiredCapabilitiesAvailable() {
...
-                    if (null != identityStoreConnectorConfig) {
-                        String connectorType = identityStoreConnectorConfig.getConnectorType();
-
-                        IdentityStoreConnectorFactory identityStoreConnectorFactory = CarbonSecurityDataHolder
-                                .getInstance().getIdentityStoreConnectorFactoryMap().get(connectorType);
+            DomainConfig domainConfig = DomainConfigBuilder.getDomainConfig(claimConfig.getMetaClaims());
+            CarbonSecurityDataHolder.getInstance().setDomainConfig(domainConfig);
+        } catch (DomainConfigException e) {
+            log.error("Error occurred in building the domain configuration - " + e.getMessage());
+        }
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityComponent.java]: pass 'e' as a different parameter

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityDataHolder.java [hunk 14]
```diff
@@ -109,11 +114,16 @@ void registerIdentityStoreConnectorFactory(String key,
...
-    void registerCacheService(CarbonCachingService carbonCachingService)  {
+    void registerCacheService(CarbonCachingService carbonCachingService) {
         this.carbonCachingService = carbonCachingService;
     }
 
     public CarbonCachingService getCarbonCachingService() {
+
+        if (this.carbonCachingService == null) {
+            throw new NullPointerException("Carbon caching service is null");
+        }
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityDataHolder.java]: Throw CarbonSecurityDataHolder Exception

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityDataHolder.java [hunk 15]
```diff
@@ -130,6 +140,11 @@ public BundleContext getBundleContext() {
     }
 
     public ClaimConfig getClaimConfig() {
+
+        if (this.claimConfig == null) {
+            throw new NullPointerException("Claim configuration is null.");
+        }
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityDataHolder.java]: Throw CarbonSecurityDataHolder Exception

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityDataHolder.java [hunk 16]
```diff
@@ -138,6 +153,11 @@ void setClaimConfig(ClaimConfig claimConfig) {
     }
 
     public DomainConfig getDomainConfig() {
+
+        if (this.domainConfig == null) {
+            throw new NullPointerException("Domain configuration is null.");
+        }
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/internal/CarbonSecurityDataHolder.java]: Throw CarbonSecurityDataHolder Exception

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/bean/Domain.java [hunk 17]
```diff
@@ -146,14 +168,49 @@ public CredentialStoreConnector getCredentialStoreConnectorFromId(String credent
...
+            for (MetaClaimMapping metaClaimMapping : metaClaimMappingList) {
+
+                if (claimURI.equals(metaClaimMapping.getMetaClaim().getClaimURI())) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/bean/Domain.java]: Convert logic to Java 8 streams

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStore.java [hunk 18]
```diff
@@ -50,12 +51,5 @@ void init(DomainManager domainManager, Map<String, CredentialStoreConnectorConfi
...
      */
-    AuthenticationContext authenticate(Callback[] callbacks) throws AuthenticationFailure;
-
-    /**
-     * Get all available credential store connector names.
-     *
-     * @return Map of display names against the connector id.
-     */
-    Map<String, String> getAllCredentialStoreNames();
+    AuthenticationContext authenticate(Callback[] callbacks) throws AuthenticationFailure, DomainException;
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStore.java]: Wrap DomainException and throw a CredentialStoreException instead

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java [hunk 19]
```diff
@@ -94,7 +86,8 @@ public void init(
     }
 
     @Override
-    public AuthenticationContext authenticate(Callback[] callbacks) throws AuthenticationFailure {
+    public AuthenticationContext authenticate(Callback[] callbacks)
+            throws AuthenticationFailure, DomainException {
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java]: Wrap DomainException and throw a CredentialStoreException instead

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java [hunk 20]
```diff
@@ -143,34 +137,40 @@ public AuthenticationContext authenticate(Callback[] callbacks) throws Authentic
...
+                return this.domainManager.getDomainFromUserName(username);
+            }
         }
 
-        return credentialStoreConnectorsMap.entrySet()
-                .stream()
-                .collect(Collectors.toMap(Map.Entry::getKey,
-                        entry -> entry.getValue().getCredentialStoreConfig().getStoreProperties()
-                                .getProperty(UserCoreConstants.USERSTORE_DISPLAY_NAME, "")));
+        throw new CredentialStoreException("NameCallBack instance not found in the callbacks array");
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/CredentialStoreImpl.java]: Wrap DomainException here

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStore.java [hunk 21]
```diff
@@ -66,7 +68,7 @@ void init(DomainManager domainManager, Map<String, IdentityStoreConnectorConfig>
      * @throws IdentityStoreException Identity Store Exception.
      * @throws UserNotFoundException  User Not Found Exception.
      */
-    User getUser(Claim claim) throws IdentityStoreException, UserNotFoundException;
+    User getUser(Claim claim) throws IdentityStoreException, UserNotFoundException, DomainException;
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStore.java]: Wrap DomainException here

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStore.java [hunk 22]
```diff
@@ -76,7 +78,7 @@ void init(DomainManager domainManager, Map<String, IdentityStoreConnectorConfig>
      * @throws IdentityStoreException Identity Store Exception.
      * @throws UserNotFoundException  User Not Found Exception.
      */
-    User getUser(Callback[] callbacks) throws IdentityStoreException, UserNotFoundException;
+    User getUser(Callback[] callbacks) throws IdentityStoreException, UserNotFoundException, DomainException;
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStore.java]: Wrap DomainException here

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java [hunk 23]
```diff
@@ -97,62 +87,67 @@ public void init(DomainManager domainManager, Map<String, IdentityStoreConnector
...
         }
-        throw userNotFoundException;
+
+        throw new UserNotFoundException("User not found for the given name.");
     }
 
-    // TODO: <VIDURA> Is it correct to throw an exception here?
     @Override
-    public User getUser(Callback[] callbacks) throws IdentityStoreException, UserNotFoundException {
+    public User getUser(Callback[] callbacks) throws IdentityStoreException, UserNotFoundException, DomainException {
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java]: Wrap DomainException here

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java [hunk 24]
```diff
@@ -97,62 +87,67 @@ public void init(DomainManager domainManager, Map<String, IdentityStoreConnector
...
+                resolveDomain(callbacks).getIdentityStoreConnectorMap();
 
         for (IdentityStoreConnector identityStoreConnector : identityStoreConnectorsMap.values()) {
             try {
-                return identityStoreConnector.getUserBuilder(callbacks)
-                        .setIdentityStore(carbonRealmService.getIdentityStore())
-                        .setAuthorizationStore(carbonRealmService.getAuthorizationStore())
-                        .setClaimManager(carbonRealmService.getClaimManager())
-                        .build();
+                // The getUserBuilder method will set a unique user id
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java]: This comment is wrong

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java [hunk 25]
```diff
@@ -195,11 +191,17 @@ public User getUser(Callback[] callbacks) throws IdentityStoreException, UserNot
             if (userCount > offset) {
                 users.addAll(identityStoreConnector.getUserBuilderList(attributeName, attributeValue, offset, length)
                         .stream()
-                        .map(userBuilder -> userBuilder
-                                .setIdentityStore(carbonRealmService.getIdentityStore())
-                                .setAuthorizationStore(carbonRealmService.getAuthorizationStore())
-                                .setClaimManager(carbonRealmService.getClaimManager())
-                                .build())
+                        .map(userBuilder -> {
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java]: Convert this to a loop because of the exception

File: components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java [hunk 26]
```diff
@@ -390,60 +415,108 @@ public Group getGroup(Claim claim) throws IdentityStoreException, GroupNotFoundE
         for (IdentityStoreConnector identityStoreConnector : identityStoreConnectorsMap.values()) {
             userList.addAll(identityStoreConnector.getUserBuildersOfGroup(groupID)
                     .stream()
-                    .map(userBuilder -> userBuilder
-                            .setIdentityStore(carbonRealmService.getIdentityStore())
-                            .setAuthorizationStore(carbonRealmService.getAuthorizationStore())
-                            .setClaimManager(carbonRealmService.getClaimManager())
-                            .build())
+                    .map(userBuilder -> {
```
Comment (vidurananayakkara) [inline components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/user/core/store/IdentityStoreImpl.java]: Remove this loop because of the exception
